aiida/orm/nodes/data/code/local.py

```diff
@@ -0,0 +1,137 @@
+# -*- coding: utf-8 -*-
+###########################################################################
+# Copyright (c), The AiiDA team. All rights reserved.                     #
+# This file is part of the AiiDA code.                                    #
+#                                                                         #
+# The code is hosted on GitHub at https://github.com/aiidateam/aiida-core #
+# For further information on the license, see the LICENSE.txt file        #
+# For further information please visit http://www.aiida.net               #
+###########################################################################
+"""Data plugin representing an executable code stored in AiiDA's storage."""
+import pathlib
+
+import click
+
+from aiida.common import exceptions
+from aiida.common.lang import type_check
+from aiida.orm import Computer
+
+from .abstract import AbstractCode
+
+__all__ = ('LocalCode',)
+
+
+class LocalCode(AbstractCode):
+    """Data plugin representing an executable code stored in AiiDA's storage."""
+
+    _KEY_ATTRIBUTE_FILEPATH_EXECUTABLE: str = 'filepath_executable'
+
+    def __init__(self, filepath_executable: str, filepath_files: pathlib.Path = None, **kwargs):
+        """Construct a new instance.
+
+        .. note:: If the files necessary for this code are not all located in a single directory or the directory
+            contains files that should not be uploaded, and so the ``filepath_files`` cannot be used. One can use the
+            methods of the :class:`aiida.orm.nodes.repository.NodeRepository` class. This can be accessed through the
+            ``base.repository`` attribute of the instance after it has been constructed. For example::
+
+                code = LocalCode(filepath_executable='some_name.exe')
+                code.put_object_from_file()
+                code.put_object_from_filelike()
+                code.put_object_from_tree()
+
+        :param filepath_executable: The relative filepath of the executable within the directory of uploaded files.
+        :param filepath_files: The filepath to the directory containing all the files of the code.
```

Comment:
> The expected behavior of the local code is still very vague to me. The executable is supposed to be one of the file in the directory `filepath_files`, if so it makes no sense not to set `filepathe_files`.


True, I will remove the `None` default from the constructor.

> If I understand correctly, `filepath_files` need to be the path to a directory of the local machine where aiida daemon runs, and it need to be a absolute path. Then, this should also be validated.

It doesn't have to be a filepath to where the AiiDA daemon runs, because when a calculation with this code is run, the content of what *was* in `filepath_files` is copied to the working directory of the remote computer. However, by that point the files will be taken from the repository. When the `PortableCode` is constructed, the contents of `filepath_files` are copied to the repository (this is done in the constructor). This is also the point where the validation is done (implicitly, by the `self.base.repository.put_object_from_tree` call).

Reply:
Makes sense to me. Thanks for the explanation!